File: app/schemas/dashboard_bundle.py
"""Composite dashboard bundle (DASH-04).

One response carrying everything the dashboard's per-chart surface needs for a
given date. The web dashboard previously issued ~13 parallel requests per chart
per date (and the family view repeated that per member); this schema is the
single-round-trip replacement. Every section is nullable: a section that fails
to compute is returned as ``null`` with a short note under ``errors`` instead
of failing the whole bundle (pairs with DASH-02's fail-soft requirement).

The individual endpoints remain in place for mobile until it migrates.
"""
from __future__ import annotations

from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.chart_explanation import ChartExplanationData
from app.schemas.charts import ChartCalculateResponseData, ChartSummaryData
from app.schemas.daily_guidance import (
    DailyGuidanceData,
    DailyGuidanceRangeData,
    WeekAheadData,
)
from app.schemas.dasha import DashaTimelineResponseData, ResponseMeta
from app.schemas.life_areas import LifeAreasResponseData
from app.schemas.panchangam import PanchangamDailyResponseData, PanchangamTimingsData
from app.schemas.peyarchi import PeyarchiEvent
from app.schemas.transits import SaniCycleData, TransitSnapshotData
from app.services.nakshatra_content import NakshatraCard


class ChartDashboardBundleData(BaseModel):
    chart_id: UUID = Field(alias="chartId")
    date_local: date = Field(alias="dateLocal")
    chart: ChartCalculateResponseData | None = None
    summary: ChartSummaryData | None = None
    daily_guidance: DailyGuidanceData | None = Field(default=None, alias="dailyGuidance")
    # 3-day window starting at date_local (today..+2) — feeds the evening preview.
    daily_guidance_range: DailyGuidanceRangeData | None = Field(default=None, alias="dailyGuidanceRange")
    # Combined maha+antar+pratyantar timeline; rows carry their `level` tag.
    dasha: DashaTimelineResponseData | None = None
    transit: TransitSnapshotData | None = None
    sani: SaniCycleData | None = None
    peyarchi_upcoming: list[PeyarchiEvent] | None = Field(default=None, alias="peyarchiUpcoming")
    explanation: ChartExplanationData | None = None
    panchangam: PanchangamDailyResponseData | None = None
    panchangam_timings: PanchangamTimingsData | None = Field(default=None, alias="panchangamTimings")
    life_areas: LifeAreasResponseData | None = Field(default=None, alias="lifeAreas")
    week_ahead: WeekAheadData | None = Field(default=None, alias="weekAhead")
    nakshatra_card: NakshatraCard | None = Field(default=None, alias="nakshatraCard")
    # Which saved location the panchangam was computed for (current wins over
    # birth), and its IANA timezone — the Today tab computes "now" in this zone
    # so horai/NOW-marker/countdowns stay location-true for diaspora (DASH-01).
    panchangam_location: Literal["current", "birth"] | None = Field(default=None, alias="panchangamLocation")
    panchangam_timezone: str | None = Field(default=None, alias="panchangamTimezone")
    # Section name -> short failure note for any section that came back null
    # because its computation raised. Purely diagnostic.
    errors: dict[str, str] = Field(default_factory=dict)

    model_config = ConfigDict(populate_by_name=True)


class ChartDashboardBundleResponse(BaseModel):
    success: bool = True
    data: ChartDashboardBundleData
    meta: ResponseMeta

    model_config = ConfigDict(populate_by_name=True)

File: packages/shared/src/api/dashboardBundle.ts
import { getApiClient } from "./client";
import type {
  ChartCalculateResponseData,
  ChartExplanationData,
  ChartSummaryData,
  DailyGuidanceData,
  DailyGuidanceRangeData,
  DashaTimelineResponseData,
  LifeAreasResponseData,
  NakshatraCardData,
  PanchangamDailyResponseData,
  PanchangamTimingsData,
  PeyarchiEvent,
  SaniCycleData,
  TransitSnapshotData,
  WeekAheadData,
} from "../types";

/**
 * One response carrying everything the dashboard renders for a chart+date —
 * replaces the ~13 parallel per-chart requests (DASH-04). Every section is
 * nullable: the backend isolates section failures and reports them under
 * `errors` instead of failing the bundle (DASH-02), so consumers must render
 * a graceful gap (or retry affordance) for null sections.
 */
export type ChartDashboardBundleData = {
  chartId: string;
  dateLocal: string;
  chart: ChartCalculateResponseData | null;
  summary: ChartSummaryData | null;
  dailyGuidance: DailyGuidanceData | null;
  dailyGuidanceRange: DailyGuidanceRangeData | null;
  /** Combined maha+antar+pratyantar timeline; rows carry their `level` tag. */
  dasha: DashaTimelineResponseData | null;
  transit: TransitSnapshotData | null;
  sani: SaniCycleData | null;
  peyarchiUpcoming: PeyarchiEvent[] | null;
  explanation: ChartExplanationData | null;
  panchangam: PanchangamDailyResponseData | null;
  panchangamTimings: PanchangamTimingsData | null;
  lifeAreas: LifeAreasResponseData | null;
  weekAhead: WeekAheadData | null;
  nakshatraCard: NakshatraCardData | null;
  /** Which saved location the panchangam used ("current" wins over "birth"). */
  panchangamLocation: "current" | "birth" | null;
  /** IANA timezone of that location — compute "now" in this zone (DASH-01). */
  panchangamTimezone: string | null;
  /** The place's own name, for "Timings for Chennai" (§2.4). `panchangamLocation`
   *  names the rule that picked it, which is not something to show a reader. */
  panchangamPlace?: string | null;
  /** When a human last vouched for the saved location — moved by a change *and*
   *  by a "Keep" confirmation. */
  locationConfirmedAt?: string | null;
  /** The §2 backstop, server-computed on LOCATION_CHECK_DUE_DAYS (R2: 45) so
   *  web and mobile cannot drift apart on the cadence. The device-timezone
   *  mismatch prompt is the main path and stays client-side — only the client
   *  knows what zone the device is in. */
  locationCheckDue?: boolean;
  /** Section name -> short failure note for sections returned as null. */
  errors: Record<string, string>;
};

export const dashboardBundleKeys = {
  bundle: (chartId: string, date: string) => ["chart", "dashboard-bundle", chartId, date] as const,
};

/**
 * Backend: GET /api/v1/charts/{chart_id}/dashboard-bundle
 * (app/api/charts.py::get_dashboard_bundle — query params `date`, `language`).
 */
export function getChartDashboardBundle(
  chartId: string,
  date: string,
  language = "ta-en",
): Promise<{ success: boolean; data: ChartDashboardBundleData }> {
  return getApiClient().get(`/charts/${chartId}/dashboard-bundle`, {
    date,
    language,
  }) as Promise<{ success: boolean; data: ChartDashboardBundleData }>;
}
